zebrad/src/components/inbound.rs

```diff
@@ -111,6 +111,73 @@ pub struct InboundSetupData {
     pub latest_chain_tip: zs::LatestChainTip,
 }
 
+/// Caches and refreshes a partial list of peer addresses to be returned as a `GetAddr` response.
+pub struct CachedPeerAddrs {
+    /// A shared list of peer addresses.
+    address_book: Arc<Mutex<zn::AddressBook>>,
+
+    /// An owned list of peer addresses used as a `GetAddr` response.
+    cached_addrs: Vec<MetaAddr>,
+
+    /// Instant after which `cached_addrs` should be refreshed.
+    refresh_time: Instant,
+}
+
+impl CachedPeerAddrs {
+    /// Creates a new [`CachedPeerAddrs`].
+    fn new(address_book: Arc<Mutex<AddressBook>>) -> Self {
+        let cached_addrs = address_book
+            .lock()
+            .expect("previous thread panicked while holding the address book lock")
+            .sanitized_window();
+
+        let refresh_time = Instant::now() + INBOUND_CACHED_ADDRS_REFRESH_INTERVAL;
+
+        Self {
+            address_book,
+            cached_addrs,
+            refresh_time,
+        }
+    }
+
+    /// Refreshes the `cached_addrs` if the time has past `refresh_time`
+    fn try_refresh(&mut self) {
+        let now = Instant::now();
+
+        // return early if the cached addresses are still fresh
+        if now < self.refresh_time {
+            return;
+        }
```

Comment:
Previously we've had network issues where we tried to do too much during startup, the inbound service queue got full, and then we disconnected some peers. So I'd like to minimise locking at startup.

Also, this code has a potential bug where it runs when the address book is empty, then keeps an empty cache for 10 minutes. This is much more likely to happen right at startup.

So I suggest making these changes:
- create an empty address book at startup
- try to refresh if the cache is stale or empty

This also simplifies the code, so we don't have to keep the sanitization and refresh intervals in `new()` and `try_refresh()` in sync.

```suggestion
    /// Creates a new empty [`CachedPeerAddrs`].
    fn new(address_book: Arc<Mutex<AddressBook>>) -> Self {
        Self {
            address_book,
            cached_addrs: Vec::new(),
            refresh_time: Instant::now(),
        }
    }

    /// Refreshes the `cached_addrs` if the time has past `refresh_time` or the cache is empty
    fn try_refresh(&mut self) {
        let now = Instant::now();

        // return early if there are some cached addresses, and they are still fresh
        if now < self.refresh_time && !self.cached_addrs.is_empty() {
            return;
        }
```